### PR TITLE
DDF-04383 Prevent users from overriding the workspace sharing settings if another user has made changes

### PR DESCRIPTION
**Pull request by mojogitoverhere**
#### What does this PR do?
Prevents users from overriding the workspace sharing settings if another user has made changes since the page was loaded. When workspace settings are saved, a request is made to fetch the latest workspace state. If the state is the same, then we allow the changes. Otherwise, we tell the user they must refresh the page and to fetch the latest changes.
#### Select relevant component teams: 
@codice/ui 
@djblue @willwill96 @gjvera @rymach 
#### Ask 2 committers to review/merge the PR and tag them here.

@andrewkfiedler
@bdeining

#### How should this be tested?

Create a new workspace and open the workspace sharing settings. Open another tab and open the workspace sharing settings for the same workspace. Change the permissions in the first tab and click `Apply`. The settings should be saved successfully. Now change the settings in the second tab and click `Apply'. The settings should fail to be saved and you should see the error message below.
#### What are the relevant tickets?
For Jira:
[](https://codice.atlassian.net/browse/)

For GH Issues:
Fixes: #4383 

#### Screenshots

screen shot 2019-02-14 at 1 18 02 pm

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx [hunk 1]
```diff
@@ -110,7 +115,46 @@ export class Sharing extends React.Component<Props, State> {
...
       },
     ]
+    this.attemptSave(attributes)
+  }
+
+  attemptSave = (attributes: any) => {
+    fetch('/search/catalog/internal/metacard/' + this.props.id)
+      .then(res => res.json())
+      .then(data => {
+        // Only allow a user to save if the previous workspace state is the same as the latest workspace state
```
Comment (gjvera) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx]: ✏️ I don't really think this comment is all that necessary. I think the if statement is pretty self explanatory.

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx [hunk 2]
```diff
@@ -110,7 +115,46 @@ export class Sharing extends React.Component<Props, State> {
...
+          JSON.stringify(data.metacards[0]) ===
+          JSON.stringify(this.state.previousWorkspace)
+        ) {
+          this.doSave(attributes)
+        } else {
+          announcement.announce(
+            {
+              title: 'The workspace settings could not be updated',
+              message:
+                'The workspace has been modified by another. Please refresh the page and reattempt your changes.',
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx]: super nit
```suggestion
                'The workspace has been modified by another user. Please refresh the page and reattempt your changes.',
```

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx [hunk 3]
```diff
@@ -110,7 +115,46 @@ export class Sharing extends React.Component<Props, State> {
...
+
+  attemptSave = (attributes: any) => {
+    fetch('/search/catalog/internal/metacard/' + this.props.id)
+      .then(res => res.json())
+      .then(data => {
+        // Only allow a user to save if the previous workspace state is the same as the latest workspace state
+        // so that changes made by other users are not clobbered.
+        // NOTE: This is a temporary workaround.
+        if (
+          JSON.stringify(data.metacards[0]) ===
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx]: Do you have access here to the attribute for last modified date? That comparison would be a little faster than the whole metacard.

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx (hunk 1 above)
Comment (mojogitoverhere) [reply ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx]: Removed

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx (hunk 3 above)
Comment (mojogitoverhere) [reply ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx]: That's a good point. I'll try it out

**Comment by mojogitoverhere** [comment]
build now

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx [hunk 4]
```diff
@@ -110,7 +115,43 @@ export class Sharing extends React.Component<Props, State> {
         values: users.filter(e => e.access === Access.Share).map(e => e.value),
       },
     ]
+    this.attemptSave(attributes)
+  }
+
+  attemptSave = (attributes: any) => {
```
Comment (samuelechu) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx]: Have you looked into using async/await?

**Comment by rymach** [comment]
> Have we considered going with a more robust solution that would work for any metacard?
> 
> For example, if we sent over the version number we're trying to modify (or possibly the last modified date) we could modify the endpoints to use this information to prevent this scenario.
> 
> Forcing a refresh isn't ideal. Could we offer to merge in the latest updates?

merging sounds really hard to get right

**Comment by rymach** [comment]
I think even confluence just warns you that someone else edited the page while you were editing... although IIRC they notify you on the page before you try to save

**Comment by rymach** [comment]
> Have we considered going with a more robust solution that would work for any metacard?
> 
> For example, if we sent over the version number we're trying to modify (or possibly the last modified date) we could modify the endpoints to use this information to prevent this scenario.

+1

**Comment by mojogitoverhere** [comment]
@andrewkfiedler I agree this solution could be more robust, but this is more of a temporary workaround for this one scenario. Even modifying the endpoints based on the version or last modified data isn't ideal - it would break in HA environments with multiple DDF nodes. According to @pklinef and @troymohl, the long term solution  is to expose solr's built in versioning feature through the framework, which is a large undertaking.

As far as merging the changes together, I agree with @rymach that it could be really hard to get right. I wanted to keep this as simple as possible but I'm open to giving it a shot though.

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx [hunk 5]
```diff
@@ -118,41 +118,43 @@ export class Sharing extends React.Component<Props, State> {
...
+    try {
+      const res = await fetch(
+        '/search/catalog/internal/metacard/' + this.props.id
+      )
+      const data = await res.json()
+      if (
+        data.metacards[0]['metacard.modified'] ===
+        this.state.previousWorkspace['metacard.modified']
+      ) {
+        debugger
```
Comment (gjvera) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx]: remove

Comment (mojogitoverhere) [reply ui/packages/catalog-ui-search/src/main/webapp/react-component/container/sharing/sharing.tsx]: whoops! 😬 removed

**Comment by mojogitoverhere** [comment]
build now

**Comment by bdeining** [comment]
build now

**Comment by rymach** [comment]
there's still a small race condition here:
User A opens sharing
User B opens sharing
User A makes changes and hits save
User B makes changes and hits save
User A's thread pulls from solr
User B's thread pulls from solr
User B's changes commited to db
User A's changes commited to db

**Comment by rymach** [comment]
IMO using the solr version field is the way to go

**Comment by mojogitoverhere** [comment]
@rymach I completely agree. However, that would involve exposing the version field through the catalog framework, which is out of scope. This ticket was intended to be a small improvement over the current behavior, which doesn't do any checking at all.

**Comment by rymach** [comment]
> @rymach I completely agree. However, that would involve exposing the version field through the catalog framework, which is out of scope. This ticket was intended to be a small improvement over the current behavior, which doesn't do any checking at all.

I would be hesitant to use a solution that introduces a race condition that will be difficult to debug in the future.

**Comment by mojogitoverhere** [comment]
Any idea how to avoid it without the full solr version solution? I considered doing all the logic on the backend by synchronizing the fetch and update for the workspace. If the workspace wasn't previously modified, I would do the update, otherwise I would return a 409 Conflict and have the user retry. But that solution won't work in an environment with multiple nodes.

**Comment by mojogitoverhere** [comment]
build now

**Comment by mojogitoverhere** [comment]
@andrewkfiedler I added a loading indicator and a comment about how this is only a temporary solution

build now

**Comment by gjvera** [comment]
hero in progress

**Comment by mojogitoverhere** [comment]
fixed bug caught during hero (thanks @gjvera!)
build now

**Comment by gjvera** [comment]
🎉 Hero Successful 🎉 
Created a few workspaces 
shared them with a few different accounts
updated the sharing settings on one account 
tried overwriting them with the other account 
saw the error and was prompted to refresh
if there was no changes made by another user sharing still worked 

LGTM!!

**Pull request merged**